### PR TITLE
Extend the Annotations framework

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR extends the current annotations framework, to:
- User VMs (existing)
- Domains (existing)
- Hosts (existing)
- Instance groups
- SSH keypairs
- Kubernetes Clusters
- Volumes
- Snapshots
- VM Snapshots
- Networks
- VPCs
- Public IP addresses
- VPN customer gateways
- Templates
- ISOs
- Service offerings
- Disks offerings
- Network offerings
- Zones
- Pods
- Clusters
- Primary Storages
- Secondary Storages
- System VMs
- Virtual Routers

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

#### New comments section
image
- The Comments section is added as a new section on the sidebar
- Users will see all their comments on different entities and can navigate to the entity by clicking on the entity ID
- Admin will see all his comments and all the comments system-wide
   - Admin has the ability to create private comments and switch its visibility

#### Quick visibility of comments on entities
image
- The listing tables are extended, showing an icon next to the entity name indicating it has comments

### How Has This Been Tested?
Tested on KVM environment (HW nor relevant) - created at least one entity/resource of each item in the list in the description. Created at least one comment per item by admin and user

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

File: api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/AddAnnotationCmd.java [hunk 1]
```diff
@@ -63,6 +69,10 @@ public String getEntityUuid() {
         return entityUuid;
     }
 
+    public boolean isAdminsOnly() {
+        return adminsOnly != null && adminsOnly;
```
Comment (GutoVeronezi) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/AddAnnotationCmd.java:73]: We could use `org.apache.commons.lang3.BooleanUtils` here:

```java
...
return BooleanUtils.toBoolean(adminsOnly);
...
```

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 2]
```diff
@@ -70,43 +88,103 @@ public AnnotationResponse addAnnotation(String text, EntityType type, String uui
...
+        if (accountVO == null) {
+            throw new CloudRuntimeException("Cannot find account with ID + " + accountId);
+        }
+        Long roleId = accountVO.getRoleId();
+        Role role = roleService.findRole(roleId);
+        if (role == null) {
+            throw new CloudRuntimeException("Cannot find role with ID " + roleId);
+        }
+        return RoleType.Admin.equals(role.getRoleType()) || RoleType.DomainAdmin.equals(role.getRoleType()) ||
+                RoleType.ResourceAdmin.equals(role.getRoleType());
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:139]: We could extract these roles to a list (constant) and verify if it contains the specified role.

**Comment by nvazquez** [comment]
Thanks @GutoVeronezi I have addressed your comments, this is still in progress so will probably ask to re-review when it's ready

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@GutoVeronezi @DaanHoogland @davidjumani please review

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by GutoVeronezi** [comment]
@nvazquez ASAP I will review this :smiley:

**Review by DaanHoogland — COMMENTED** [review]
code looks good, but serious tests needed (unit or integration). nice work on the code/refactor.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/UpdateAnnotationVisibilityCmd.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    public void execute() throws ResourceUnavailableException, InsufficientCapacityException, ServerApiException,
+            ConcurrentOperationException, ResourceAllocationException, NetworkRuleConflictException {
+        AnnotationResponse annotationResponse = annotationService.updateAnnotationVisibility(this);
+        annotationResponse.setResponseName(getCommandName());
+        this.setResponseObject(annotationResponse);
+    }
+
+    @Override
+    public String getCommandName() {
+        return APINAME.toLowerCase() + BaseCmd.RESPONSE_SUFFIX;
```
Comment (GutoVeronezi) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/UpdateAnnotationVisibilityCmd.java:67]: We can use String.format to concatenate these strings.
```suggestion
        return String.format("%s%s", APINAME.toLowerCase(), BaseCmd.RESPONSE_SUFFIX);
```

File: api/src/main/java/org/apache/cloudstack/api/command/admin/host/UpdateHostCmd.java [hunk 4]
```diff
@@ -124,7 +124,7 @@ public void execute() {
         try {
             result = _resourceService.updateHost(this);
             if(getAnnotation() != null) {
-                annotationService.addAnnotation(getAnnotation(), AnnotationService.EntityType.HOST, result.getUuid());
+                annotationService.addAnnotation(getAnnotation(), AnnotationService.EntityType.HOST, result.getUuid(),true);
```
Comment (GutoVeronezi) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/host/UpdateHostCmd.java:127]:
```suggestion
                annotationService.addAnnotation(getAnnotation(), AnnotationService.EntityType.HOST, result.getUuid(), true);
```

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41510to41600.java [hunk 5]
```diff
@@ -64,6 +64,32 @@ public boolean supportsRollingUpgrade() {
...
+            if (!rs.isClosed())  {
+                rs.close();
+            }
+            if (!pstmt.isClosed())  {
+                pstmt.close();
+            }
+            LOG.debug("Successfully generated uuid for existing ssh key-pairs");
+        } catch (SQLException e) {
+            String errMsg = "Exception while generating uuid for existing ssh key-pairs: " + e.getMessage();
+            LOG.error(errMsg);
```
Comment (GutoVeronezi) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41510to41600.java:90]: We could pass the exception as parameter here.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 6]
```diff
@@ -53,64 +113,198 @@
...
 
-        AnnotationVO annotation = new AnnotationVO(text, type, uuid);
+        AnnotationVO annotation = new AnnotationVO(text, type, uuid, adminsOnly);
         annotation.setUserUuid(userUuid);
         annotation = annotationDao.persist(annotation);
         return createAnnotationResponse(annotation);
     }
 
+    private void checkAnnotationPermissions(String entityUuid, EntityType type, UserVO user) {
+        if (!isCallingUserAdmin()) {
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:132]: We could invert this first if and add a return (and maybe a log), to reduce code indentation.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 7]
```diff
@@ -53,64 +113,198 @@
...
+            if (!type.isUserAllowed()) {
+                throw new CloudRuntimeException("User " + user.getUsername() + " is not allowed to add annotations on type " + type.name());
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:134]: We could use `String.format` here.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 8]
```diff
@@ -53,64 +113,198 @@
...
+    }
+
     @Override
     @ActionEvent(eventType = EventTypes.EVENT_ANNOTATION_REMOVE, eventDescription = "removing an annotation on an entity")
     public AnnotationResponse removeAnnotation(RemoveAnnotationCmd removeAnnotationCmd) {
         String uuid = removeAnnotationCmd.getUuid();
-        if(LOGGER.isDebugEnabled()) {
-            LOGGER.debug("marking annotation removed: " + uuid);
+        AnnotationVO annotation = annotationDao.findByUuid(uuid);
+        if (isCallingUserAllowedToRemoveAnnotation(annotation)) {
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:145]: We could invert this if to reduce code indentation.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 9]
```diff
@@ -53,64 +113,198 @@
...
+        return createAnnotationResponse(annotation);
+    }
+
+    @Override
+    public AnnotationResponse updateAnnotationVisibility(UpdateAnnotationVisibilityCmd cmd) {
+        String uuid = cmd.getUuid();
+        Boolean adminsOnly = cmd.getAdminsOnly();
         AnnotationVO annotation = annotationDao.findByUuid(uuid);
-        annotationDao.remove(annotation.getId());
+        if (annotation != null && isCallingUserAdmin()) {
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:162]: We could invert this if to reduce code indentation.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 10]
```diff
@@ -53,64 +113,198 @@
...
     @Override
     @ActionEvent(eventType = EventTypes.EVENT_ANNOTATION_REMOVE, eventDescription = "removing an annotation on an entity")
     public AnnotationResponse removeAnnotation(RemoveAnnotationCmd removeAnnotationCmd) {
         String uuid = removeAnnotationCmd.getUuid();
-        if(LOGGER.isDebugEnabled()) {
-            LOGGER.debug("marking annotation removed: " + uuid);
+        AnnotationVO annotation = annotationDao.findByUuid(uuid);
+        if (isCallingUserAllowedToRemoveAnnotation(annotation)) {
+            if(LOGGER.isDebugEnabled()) {
+                LOGGER.debug("marking annotation removed: " + uuid);
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:147]: We could improve this log by adding more context info.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 11]
```diff
@@ -53,64 +113,198 @@
...
+
+    @Override
+    public AnnotationResponse updateAnnotationVisibility(UpdateAnnotationVisibilityCmd cmd) {
+        String uuid = cmd.getUuid();
+        Boolean adminsOnly = cmd.getAdminsOnly();
         AnnotationVO annotation = annotationDao.findByUuid(uuid);
-        annotationDao.remove(annotation.getId());
+        if (annotation != null && isCallingUserAdmin()) {
+            if(LOGGER.isDebugEnabled()) {
+                LOGGER.debug("updating annotation visibility: " + uuid);
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:164]: We could improve this log by adding more context info.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 12]
```diff
@@ -53,64 +113,198 @@
...
-        if(LOGGER.isDebugEnabled()) {
-            LOGGER.debug("marking annotation removed: " + uuid);
+        AnnotationVO annotation = annotationDao.findByUuid(uuid);
+        if (isCallingUserAllowedToRemoveAnnotation(annotation)) {
+            if(LOGGER.isDebugEnabled()) {
+                LOGGER.debug("marking annotation removed: " + uuid);
+            }
+            annotationDao.remove(annotation.getId());
+        } else {
+            throw new CloudRuntimeException("Only administrators or entity owner users can delete annotations, cannot remove annotation: " + uuid);
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:151]: We could improve this exception by adding more context info.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 13]
```diff
@@ -53,64 +113,198 @@
...
         AnnotationVO annotation = annotationDao.findByUuid(uuid);
-        annotationDao.remove(annotation.getId());
+        if (annotation != null && isCallingUserAdmin()) {
+            if(LOGGER.isDebugEnabled()) {
+                LOGGER.debug("updating annotation visibility: " + uuid);
+            }
+            annotation.setAdminsOnly(adminsOnly);
+            annotationDao.update(annotation.getId(), annotation);
+        } else {
+            throw new CloudRuntimeException("Cannot update visibility for annotation: " + uuid);
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:169]: We could improve this exception by adding more context info.

File: server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java [hunk 14]
```diff
@@ -53,64 +113,198 @@
...
+
     private List<AnnotationVO> getAnnotationsForApiCmd(ListAnnotationsCmd cmd) {
         List<AnnotationVO> annotations;
-        if(cmd.getUuid() != null) {
+        String userUuid = cmd.getUserUuid();
+        String entityUuid = cmd.getEntityUuid();
+        String entityType = cmd.getEntityType();
+        String annotationFilter = isNotBlank(cmd.getAnnotationFilter()) ? cmd.getAnnotationFilter() : "all";
+        boolean isCallerAdmin = isCallingUserAdmin();
+        if ((isBlank(entityUuid) || isBlank(entityType)) && !isCallerAdmin && annotationFilter.equalsIgnoreCase("all")) {
```
Comment (GutoVeronezi) [inline server/src/main/java/org/apache/cloudstack/annotation/AnnotationManagerImpl.java:218]: `StringUtils` has the method `isAnyEmpty`, where we can validate several strings at the same time.

**Review by GutoVeronezi — COMMENTED** [review]
@nvazquez nice work. I raised some minor improvements.

**Comment by nvazquez** [comment]
Thanks @GutoVeronezi - comments addressed
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test keepEnv

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/AddAnnotationCmd.java [hunk 15]
```diff
@@ -40,11 +41,17 @@
...
     private String annotation;
+
     @Parameter(name = ApiConstants.ENTITY_TYPE, type = CommandType.STRING, description = "the entity type (only HOST is allowed atm)")
     private String entityType;
+
     @Parameter(name = ApiConstants.ENTITY_ID, type = CommandType.STRING, description = "the id of the entity to annotate")
     private String entityUuid;
 
+    @Parameter(name = ApiConstants.ADMINS_ONLY, type = CommandType.BOOLEAN,
+            description = "the annotation is visible for admins only")
```
Comment (davidjumani) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/AddAnnotationCmd.java:52]:
```suggestion
            description = "the annotation is visible for admins only. Default is true")
```

File: ui/src/components/view/AnnotationsTab.vue [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+        case 'ServiceOffering': return 'SERVICE_OFFERING'
+        case 'DiskOffering': return 'DISK_OFFERING'
+        case 'NetworkOffering': return 'NETWORK_OFFERING'
+        case 'Zone': return 'ZONE'
+        case 'Pod': return 'POD'
+        case 'Cluster': return 'CLUSTER'
+        case 'PrimaryStorage': return 'PRIMARY_STORAGE'
+        case 'SecondaryStorage': return 'SECONDARY_STORAGE'
+        case 'SystemVm': return 'SYSTEM_VM'
+        case 'VirtualRouter': return 'VR'
```
Comment (davidjumani) [inline ui/src/components/view/AnnotationsTab.vue:163]:
```suggestion
        case 'VirtualRouter': return 'VR'
        case 'Kubernetes: return ''KUBERNETES_CLUSTER'
```

File: api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/ListAnnotationsCmd.java [hunk 17]
```diff
@@ -41,11 +41,25 @@
...
+    @Parameter(name = ApiConstants.USER_ID, type = CommandType.STRING,
+            description = "optional: the id of the user of the annotation", required = false)
+    private String userUuid;
+
+    @Parameter(name = ApiConstants.ANNOTATION_FILTER,
+            type = CommandType.STRING,
+            required = true,
+            description = "possible values are \"self\" and \"all\". "
+                    + "* self : annotations that have been created by the calling user. "
+                    + "* all : all templates (only usable by admins).")
```
Comment (davidjumani) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/ListAnnotationsCmd.java:60]:
```suggestion
                    + "* all : all annotations (only usable by admins).")
```

File: api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/AddAnnotationCmd.java (hunk 15 above)
Comment (davidjumani) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/AddAnnotationCmd.java:52]: My bad

File: api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/ListAnnotationsCmd.java (hunk 17 above)
Comment (nvazquez) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/annotation/ListAnnotationsCmd.java]: Done, thanks

**Comment by nvazquez** [comment]
@blueorangutan ui

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by RodrigoDLopez — CHANGES_REQUESTED** [review]
@nvazquez great enhancement.

I'm asking for some minor changes, it's not relevant.
But, I think since your are adding new parameters to the API, we need to set the 'since' flag.

**Comment by nvazquez** [comment]
Thanks for reviewing @RodrigoDLopez, I'll work on addressing your changes

**Comment by nvazquez** [comment]
Comments addressed @RodrigoDLopez thanks! Yes, this PR needs manual testing, btw my colleagues @borisstoyanov @vladimirpetrov are planning to test it soon

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by RodrigoDLopez — APPROVED** [review]

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: test/integration/smoke/test_annotations.py [hunk 18]
```diff
@@ -31,15 +31,69 @@
...
+        cls.host = list_hosts(cls.apiclient,
+                               zoneid=cls.zone.id,
+                               type='Routing')[0]
+
+    @classmethod
+    def tearDownClass(cls):
+        try:
+            cleanup_resources(cls.apiclient, cls._cleanup)
+        except Exception as e:
+            raise Exception("Warning: Exception during cleanup : %s" % e)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_annotations.py:92]:
```suggestion
        super(TestAnnotations, cls).tearDownClass()
```

the same (or similar) should be done for `tearDown(self)`

Comment (nvazquez) [reply test/integration/smoke/test_annotations.py]: Thanks, fixed

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@GutoVeronezi @davidjumani all your remarks/concerns are met?

**Comment by nvazquez** [comment]
Marked as a draft while the manual testing by @borisstoyanov is still in process

**Review by GutoVeronezi — APPROVED** [review]
Code LGTM, however I didn't test it.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested the feature with different entities and accounts

**Comment by borisstoyanov** [comment]
let's just wait for final test run and any objections

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Pull request merged**